### PR TITLE
[improve] [broker] Let the producer request success at the first time if the previous one is inactive

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation

See https://github.com/apache/pulsar/pull/21155#discussion_r1328892859

If a producer establishes a new connection when it is reconnecting, while the previous connection is now inactive, the initial request made on the new connection will fail.  This failure will trigger the topic of cleaning up the inactive producers.  However, upon making a second request, the producer will be able to successfully establish a connection and proceed with the operation.

### Modifications

Make the initial request made on the new connection success.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: x

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java [hunk 1]
```diff
@@ -974,31 +975,40 @@ protected void internalAddProducer(Producer producer) throws BrokerServiceExcept
...
             // available. The producers related the connection that not available are automatically cleaned up.
             if (!Objects.equals(oldProducer.getCnx(), newProducer.getCnx())) {
-                oldProducer.getCnx().checkConnectionLiveness();
+                return oldProducer.getCnx().checkConnectionLiveness().thenCompose(previousIsActive -> {
+                    if (previousIsActive) {
+                        return CompletableFuture.failedFuture(new BrokerServiceException.NamingException(
+                                "Producer with name '" + newProducer.getProducerName()
+                                        + "' is already connected to topic"));
+                    } else {
+                        return internalAddProducer(newProducer);
```
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:1006]: And it also should be BUG before, right?

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:1012]: > It's better to add a comment here to explain why we should take a recursive call.

Done

> And it also should be BUG before, right?

Yes, the issue is that the producer creation will fail if a same-name producer has already closed on the client side.

**Review by liangyepianzhou — APPROVED** [review]

**Comment by AnonHxy** [comment]
There are conflicts when cherry pick this patch to branch-3.1, would you please help do it @poorbarcode

**Comment by poorbarcode** [comment]
> There are conflicts when cherry pick this patch to branch-3.1, would you please help do it @poorbarcode

Done

**Pull request merged**